signer/src/bitcoin/utxo.rs

```diff
@@ -1068,6 +1080,142 @@ impl BitcoinTxInfo {
     }
 }
 
+/// An output used as an input into a transaction, a previous output.
+#[derive(Copy, Clone, Debug)]
+pub struct PrevoutRef<'a> {
+    /// The amount locked but the output
+    pub amount: Amount,
+    /// The `scriptPubKey` locking the output
+    pub script_pubkey: &'a ScriptBuf,
+    /// The ID of the transaction that created the output.
+    pub txid: &'a Txid,
+    /// The index of the output in the transactions outputs.
+    pub output_index: u32,
+}
+
+/// A trait for deconstructing a bitcoin transaction related to the signers
+/// into its inputs and outputs.
+pub trait TxDeconstructor: InputsOutputs {
+    /// Returns a prevout given the input index.
+    ///
+    /// This function must return `Some(_)` for each `index` where
+    /// `self.inputs().get(index)` returns `Some(_)`, and must be `None`
+    /// otherwise.
+    fn prevout(&self, index: usize) -> Option<PrevoutRef>;
+
+    /// Return all inputs in this transaction if it is an sBTC transaction.
+    ///
+    /// This function returns an empty vector if it was not generated by
+    /// the signers, where the signers are identified by their
+    /// `signer_script_pubkeys`.
+    fn to_inputs(&self, signer_script_pubkeys: &HashSet<ScriptBuf>) -> Vec<TxPrevout> {
+        // If someone else created this transaction then we are not a party
+        // to any of the inputs, so we can exit early.
+        if !self.is_signer_created(signer_script_pubkeys) {
+            return Vec::new();
+        };
+
+        // This is a transaction that the signers have created. It follows
+        // a layout described in the description of `UnsignedTransaction`.
+        self.inputs()
+            .iter()
+            .enumerate()
+            .filter_map(|(index, _)| match index {
+                0 => self.vin_to_prevout(index, TxPrevoutType::SignersInput),
+                _ => self.vin_to_prevout(index, TxPrevoutType::Deposit),
+            })
+            .collect()
+    }
+
+    /// Return all outputs in this transaction that are related to the
+    /// signers.
+    ///
+    /// This function returns all outputs if the transaction is an
+    /// sBTC transaction, and only outputs that the signers can sign for
+    /// otherwise.
+    fn to_outputs(&self, signer_script_pubkeys: &HashSet<ScriptBuf>) -> Vec<TxOutput> {
+        // This transaction might not be related to the signers at all. If
+        // not then we can exit early.
+        let mut outputs = self.outputs().iter();
+        if !outputs.any(|tx_out| signer_script_pubkeys.contains(&tx_out.script_pubkey)) {
+            return Vec::new();
+        }
+
+        // If the signers did not create this transaction, but the signers
+        // control at least one output then the outputs that the signers
+        // control are donations. So we scan the outputs and exit early.
```

Comment:
maybe nit: for completeness we could mention that those cannot be deposits since deposits outputs are not "directly" controlled by signers aggregate key